hw-model/src/openocd/openocd_server.rs

```diff
@@ -0,0 +1,228 @@
+// Licensed under the Apache-2.0 license
+//
+// Derived from OpenTitan's opentitanlib with original copyright:
+//
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::io::{BufRead, BufReader, Write};
+use std::net::TcpStream;
+use std::os::unix::process::CommandExt;
+use std::path::Path;
+use std::process::{Child, Command, Stdio};
+use std::time::{Duration, Instant};
+
+use anyhow::{bail, Context, Result};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::openocd::printer;
+
+/// Errors related to the OpenOCD server.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum OpenOcdError {
+    #[error("OpenOCD initialization failed: {0}")]
+    InitializeFailure(String),
+    #[error("OpenOCD server exists prematurely")]
+    PrematureExit,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+
+/// Represents an OpenOCD server that we can interact with.
+pub struct OpenOcdServer {
+    /// OpenOCD child process.
+    server_process: Child,
+    /// Receiving side of the stream to the telnet interface of OpenOCD.
+    reader: BufReader<TcpStream>,
+    /// Sending side of the stream to the telnet interface of OpenOCD.
+    writer: TcpStream,
+}
+
+impl Drop for OpenOcdServer {
+    fn drop(&mut self) {
+        let _ = self.server_process.kill();
```

Comment:
Should this use `self.shutdown()`, or does it not need to end gracefully?

Reply:
good point; fixed